blocks/search.py

```diff
@@ -0,0 +1,280 @@
+        Computational graph which contains `sequence_generator`.
+
+    See Also
+    --------
+    :class:`SequenceGenerator`, :class:`SequenceContentAttention`.
+
+    Notes
+    -----
+    Sequence generator should use an emitter which has `probs` method and
+    one of its outputs is called `probs` e.g. :class:`SoftmaxEmitter`.
+
+    """
+    def __init__(self, beam_size, sequence_generator, comp_graph):
+        self.beam_size = beam_size
+        self.sequence_generator = sequence_generator
+        self.generate_names = sequence_generator.generate.states
+        self.context_names = sequence_generator.generate.contexts
+        self.init_computer = None
+        self.next_state_computer = None
+        self.context_computer = None
+        self.initial_state_computer = None
+        self.state_names = (sequence_generator.state_names +
+                            sequence_generator.glimpse_names +
+                            ['outputs'])
+        self.context_names = sequence_generator.context_names
+        self.comp_graph = comp_graph
+        self.inputs = OrderedDict(comp_graph.dict_of_inputs())
+        self.need_input_states = []
+        self.compiled = False
+
+    def compile_context_computer(self, contexts):
+        """Compiles `context_computer`."""
+        self.context_computer = function(list(self.inputs.values()),
+                                         list(contexts.values()),
+                                         on_unused_input='ignore')
+
+    def compile_initial_state_computer(self, generator, contexts):
+        """Compiles `initial_state_computer`."""
+        initial_states = []
+        for name in self.state_names:
+            initial_states.append(generator.initial_state(
+                name,
+                self.beam_size,
+                **contexts))
+        self.initial_state_computer = function(list(contexts.values()),
+                                               initial_states,
+                                               on_unused_input='ignore')
+
+    def compile_next_state_computer(self, generator, contexts, inner_cg):
+        """Compiles `next_state_computer`."""
+        states = []
+        for name in generator.generate.states:
+            var = VariableFilter(bricks=[generator], name='^' + name + '$',
+                                 roles=[INPUT])(inner_cg)[-1:]
+            if var:
+                self.need_input_states.append(name)
+            states.extend(var)
+
+        next_states = []
+        for name in generator.generate.states:
+            var = VariableFilter(bricks=[generator], name='^' + name + '$',
+                                 roles=[OUTPUT])(inner_cg)[-1:]
+            next_states.extend(var)
+
+        next_probs = VariableFilter(
+            bricks=[generator.readout.emitter],
+            name='^probs')(inner_cg)[-1]
+        # Create theano function for next states
+        self.next_state_computer = function(list(contexts.values()) + states,
```

Comment:
As I have already said before, I believe you have an issue here: you are not overriding the feedback to the network here. Effectively you still some random generation inside, and I have a proof. The quote below shows how your implementation assigns different scores to identical sequences, which is a clear sign of something non-deterministic happening under the hood.

```
Enter a sentence
das ist ein problem .
Encoder input: [42, 3, 0, 18, 41, 8, 18, 19, 41, 4, 8, 13, 41, 15, 17, 14, 1, 11, 4, 12, 41, 37, 43]
Target:  [42, 18, 0, 3, 41, 19, 18, 8, 41, 13, 8, 4, 41, 12, 4, 11, 1, 14, 17, 15, 41, 37, 43]
(0.915070145559)<S>sad tsi nie melborp .</S> CORRECT!
(0.913463181202)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
(0.913414438598)<S>sad tsi nie melborp .</S> CORRECT!
```